event_model/__init__.py

```diff
@@ -80,47 +80,79 @@ def _dispatch(self, name, doc, validate):
         Optionally validate that the result is still a valid document.
         """
         output_doc = getattr(self, name)(doc)
+
+        # If 'event' is not defined by the subclass but 'event_page' is, or
+        # vice versa, use that. And the same for 'datum_page' / 'datum.
+        if output_doc is NotImplemented:
+            if name == 'event':
+                event_page = pack_event_page(doc)
+                # Subclass' implementation of event_page may return a valid
+                # EventPage or None or NotImplemented.
+                output_event_page = self.event_page(event_page) or event_page
+                if output_event_page is not NotImplemented:
+                    output_doc, = unpack_event_page(output_event_page)
+            elif name == 'datum':
+                datum_page = pack_datum_page(doc)
+                # Subclass' implementation of datum_page may return a valid
+                # DatumPage or None or NotImplemented.
+                output_datum_page = self.datum_page(datum_page) or datum_page
+                if output_datum_page is not NotImplemented:
+                    if output_datum_page is None:
+                        output_datum_page = doc
+                    output_doc, = unpack_datum_page(output_datum_page)
+            elif name == 'event_page':
+                output_events = []
+                for event in unpack_event_page(doc):
+                    # Subclass' implementation of event may return a valid
+                    # Event or None or NotImplemented.
+                    output_event = self.event(event) or event
+                    if output_event is NotImplemented:
+                        break
```

Comment:
What happens if `output_event` is None?

Reply:
This line

```py
output_event = self.event(event) or event
```

ensures that if `self.event(event)` returns `None` (which is quite possible) then `output_event` will be set to `event`, which is the document originally passed in to the DocumentRouter. We assume that this is a valid Event document, never `None`. (And, if validation is turned on, we verify that assumption.)